test/TestingStrategy.md

```diff
@@ -0,0 +1,69 @@
+This document describe the testing strategy we use for all Vitess components, and the progression in scope / complexity.
+
+## Unit Tests
+
+All code should have great unit test coverage, no exception.
+
+However, since we use other components extensively, and multiple languages, it is not sufficient. Sometimes, we just gotta have a real MySQL process, if we're testing interactions with MySQL.
+
+## Client Library Tests and vtgateclienttest
+
+All client libraries should be tested with vtgateclienttest. It is meant to validate all client inputs, and make sure all corner cases are handled properly in the client. For instance, to make sure all errors the server can return are translated into the right form on the client.
+
+It is a single stand-alone process, light-weight enough that it runs very quickly and doesn't take many resources at all, and just requires a port to listen on.
+
+(note the go version is run in-process, doesn't even fork vtgateclienttest).
+
+## vttest Library, run\_local\_database.py
+
+This set of helpers has two purposes:
+
+* bring up a very light well cluster to help test client applications. YouTube uses this internally with thousands of tests.
+
+* help with Vitess unit tests when a real instance is needed (for instance, to test the interactions with MySQL are what we expect).
+
+It is expected that Vitess users may change py/vttest/environment.py to match their setup. For instance, YouTube internally does is to use different scripts to bring up a MySQL instance.
+
+These tests need to be as light weight as possible, so the developers who use Vitess can run as many unit tests as they want with minimal resources. We are currently running a single MySQL, multiple vtocc, and one vtgate. The plan is to switch to vtcombo to replace vtgate+vtocc processes with a single process. The MySQL process has multiple databases, one per keyspace / shard. It is the smallest setup with a real MySQL.
+
+This framework supports an initial schema to apply to all keyspaces / shards, per keyspace. This is also the base for testing schema changes before rolling them out to production.
+
+Due to its constant nature, this is not an appropriate framework to test cluster events, like re-sharding, re-parenting, advanced query routing, ... No replication is setup on the single MySQL instance, obviously.
+
+## Integration Tests
```

Comment:
I agree with name change, these are end to end tests. Will add expectations too.